virtual_rainforest/core/config.py

```diff
@@ -0,0 +1,254 @@
+"""The `core.config` module.
+
+The `core.config` module is used to read in the various configuration files, validate
+their contents, and then configure a ready to run instance of the virtual rainforest
+model.
+"""
+# TODO - find config folder based on command line argument
+
+import os
+import sys
+from typing import Callable
+
+import dpath.util
+import jsonschema
+import tomli_w
+
+from virtual_rainforest.core.logger import LOGGER
+
+if sys.version_info[:2] >= (3, 11):
+    import tomllib
+else:
+    import tomli as tomllib
+
+# Dictionary to store schema registry
+SCHEMA_REGISTRY: dict = {}
+
+
+def register_schema(module_name: str) -> Callable:
+    """Decorator function to add configuration schema to the registry."""
+
+    def wrap(func: Callable):
+        SCHEMA_REGISTRY[module_name] = func()
```

Comment:
Or overwrite existing `module name` and warn? That's what I'm currently using in the grid registry.

Reply:
I also think this would be a good place to validate the schema itself:

https://python-jsonschema.readthedocs.io/en/stable/api/jsonschema/protocols/#jsonschema.protocols.Validator.check_schema

I don't know whether we use a specific versioned validator or just the default, but not adding malformed schemas to the registry seems like a good built in developer safeguard.

Reply:
I would probably lean towards using a critical error rather than a warning here. Basically think in cases where a `module name` is repeated non-sensical errors downstream are all but guaranteed if the configuration process is continued.